### ISSUE TITLE
Hot Reload schema/config not working when running in kubernetes as a configmap 

### ISSUE THREAD

**Issue by kmcrawford** [apollographql/router#1695]
**Describe the bug**
I am running a router in a kubernetes deployment using `ghcr.io/apollographql/router:v1.0.0-alpha.0`. I have the supergraph and the config as a mounted volume from a configmap.

```
env:
      - name: APOLLO_ROUTER_CONFIG_PATH
        value: /app/configuration.yaml
      - name: APOLLO_ROUTER_HOT_RELOAD
        value: "true"
      - name: APOLLO_ROUTER_SUPERGRAPH_PATH
        value: /schema/supergraph.graphql
...
          volumeMounts:
            - name: router-configuration
              mountPath: /app/
              readOnly: true
            - name: router-schema
              mountPath: /schema/
              readOnly: true
      volumes:
        - name: router-configuration
          configMap:
            name: router-config
        - name: router-schema
          configMap:
            name: router-schema
```

**To Reproduce**
If I modify the configmap I expect to see `INFO apollo_router::state_machine: reloading schema` but I do not see anything in the pod logs.  I can `kubectl debug` to the pod and verify that the schema has the updated changes but it does not reload the schema.  I have verified that this works locally using local files, but does not work correctly in kubernetes.

**Expected behavior**
I expect to see `INFO apollo_router::state_machine: reloading schema`

**Comment by garypen** [comment]
I imagine this will be fixed when #1693 is available.

**Comment by kmcrawford** [comment]
@garypen this fix is on deployment to restart the router.  This ticket is about hot reload, which is without a deployment.

**Comment by garypen** [comment]
Ok. I'll update my comment.

**Comment by garypen** [comment]
I wonder if this is related to: #1476 ? I have observed that the router hot-reload doesn't work correctly with a variety of virtual filesystems. Are there any errors in the router logs near startup indicating watch issues?

**Comment by kmcrawford** [comment]
There are no errors in the debug level logs.  This error in #1476 looks to be on startup.  Everything starts fine, just doesn’t reload when the configmap is updated.

**Comment by kmcrawford** [comment]
I have verified that the configmap does get updated inside the pod. So it is doing what kubernetes states it is supposed to be doing. https://kubernetes.io/docs/concepts/configuration/configmap/#mounted-configmaps-are-updated-automatically

**Comment by kmcrawford** [comment]
This could be related to the way the watch is working in the code. Since the write event happens externally from the os perhaps it can’t catch a write event. https://github.com/apollographql/router/blob/0a94e20425241de1eb8dd93b77813844b62e6a24/apollo-router/src/files.rs#L34

**Comment by kmcrawford** [comment]
According to Hotwatch is it a wrapper for notify (https://docs.rs/hotwatch/0.2.1/hotwatch/index.html). Notify states it does have issues with fsevents for files it does not own.  I am not sure if this is the issue, but perhaps a clue in the right direction.
https://github.com/notify-rs/notify#fsevents

**Comment by kmcrawford** [comment]
I found if I delete the configmap I do see this in the logs. `Remove` isn't an event that is currently being watched, but also I wouldn't be doing an delete, but a replace that doesn't seem to fire an event.
```
"level":"DEBUG","fields":{"message":"file watcher: received event: Remove(\"/schema/..2022_09_03_12_38_50.1343268253/supergraph.graphql\")"},"target":"apollo_router::files"}
```

**Comment by kmcrawford** [comment]
I tried to create a sidecar container with a shared `emptyDir` volume with the router to copy from the configmap location to the shared volume every 5 seconds, but the router also can't see file events on the `emptyDir`

```
          - name: copy-schema
          image: alpine:latest
          command: [ "/bin/sh", "-c", "--" ]
          args: [ "while true; do cp /schema/supergraph.graphql /schema-router/supergraph.graphql;  done;" ]
          volumeMounts:
            #configmap volume
            - name: router-schema
              mountPath: /schema/
            #emptyDir volume
            - name: schema-router-configuration
              mountPath: "/schema-router"
```

**Comment by garypen** [comment]
@kmcrawford 

> I have verified that the configmap does get updated inside the pod. So it is doing what kubernetes states it is supposed to be doing. https://kubernetes.io/docs/concepts/configuration/configmap/#mounted-configmaps-are-updated-automatically

I'm looking into this a bit further and I've raised a PR with a fix for #1476. However, I recalled that there is [a limitation of kubernetes ConfigMaps](https://kubernetes.io/docs/concepts/configuration/configmap/#mounted-configmaps-are-updated-automatically) which means that changes to a configMap which makes use of a subPath (as our template does) won't be seen in a container. That makes me wonder how you managed to get your configmap to update in your pod. Can you share the details?

**Comment by garypen** [comment]
Ok. I manually verified this worked on a k3d cluster running on my laptop. I would observe that it was really slow reacting to the change, but that might be part of the complexity of k3d, docker etc rather than anything related to our solution.

**Comment by kmcrawford** [comment]
> @kmcrawford
> 
> > I have verified that the configmap does get updated inside the pod. So it is doing what kubernetes states it is supposed to be doing. https://kubernetes.io/docs/concepts/configuration/configmap/#mounted-configmaps-are-updated-automatically
> 
> I'm looking into this a bit further and I've raised a PR with a fix for #1476. It's difficult for me to test that this works in kubernetes using our own Helm Template since we use a subPath for the configuration and that won't be updated on changes.
> 
> I've verified that the changes are working correctly in Docker, so I'm reasonably confident this should work with kubernetes as well. I'll try to find some time to manually verify this using a volume mount as you are doing.

Thank you @garypen I'm aware of the limitation of subpath and am not using that in my configuration.

**Comment by garypen** [comment]
@kmcrawford  I took a look at our helm chart and concluded that we didn't need the `subPath`, so I have updated my PR to remove `subPath` from there. Thanks for filing the issue and persevering with your investigation.
